akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/HostConnectionPool.scala

```diff
@@ -0,0 +1,335 @@
+          else
+            outBuffer.addLast(resCtx)
+
+        def dispatchRequest(req: RequestContext): Unit = {
+          val slot =
+            slots.find(_.isIdle)
+              .getOrElse(throw new IllegalStateException("Tried to dispatch request when no slot is idle"))
+
+          slot.debug("Dispatching request") // FIXME: add abbreviation
+          slot.dispatchRequest(req)
+        }
+
+        private class Slot(val slotId: Int) extends SlotContext {
+          private[this] var state: SlotState = SlotState.Unconnected
+          private[this] var connection: SlotConnection = _
+          def isIdle: Boolean = state.isIdle
+          def isConnected: Boolean = state.isConnected
+          def shutdown(): Unit = {
+            // TODO: should we offer errors to the connection?
+            closeConnection()
+
+            state.onShutdown(this)
+          }
+
+          def initialize(): Unit =
+            if (slotId < settings.minConnections)
+              updateState(_.onPreConnect(this))
+
+          def onConnected(outgoing: Http.OutgoingConnection): Unit =
+            updateState(_.onConnectedAttemptSucceeded(this, outgoing))
+
+          def onConnectFailed(cause: Throwable): Unit =
+            updateState(_.onConnectionAttemptFailed(this, cause))
+
+          def dispatchRequest(req: RequestContext): Unit =
+            updateState(_.onNewRequest(this, req))
+
+          def onRequestEntityCompleted(): Unit =
+            updateState(_.onRequestEntityCompleted(this))
+          def onRequestEntityFailed(cause: Throwable): Unit =
+            updateState(_.onRequestEntityFailed(this, cause))
+
+          def onResponseReceived(response: HttpResponse): Unit =
+            updateState(_.onResponseReceived(this, response))
+
+          def onResponseEntityCompleted(): Unit =
+            updateState(_.onResponseEntityCompleted(this))
+          def onResponseEntityFailed(cause: Throwable): Unit =
+            updateState(_.onResponseEntityFailed(this, cause))
+
+          def onConnectionCompleted(): Unit = updateState(_.onConnectionCompleted(this))
+          def onConnectionFailed(cause: Throwable): Unit = updateState(_.onConnectionFailed(this, cause))
+
+          protected def updateState(f: SlotState ⇒ SlotState): Unit = {
+            val previousState = state
+            state = f(state)
+
+            if (!previousState.isIdle && state.isIdle) {
+              debug("Slot became idle... Trying to pull")
+              pullIfNeeded()
+            }
+
+            // put additional bookkeeping here (like keeping track of idle connections)
+          }
+          protected def setState(newState: SlotState): Unit =
+            updateState(_ ⇒ newState)
+
+          def debug(msg: String): Unit =
+            log.debug("[{} ({})] {}", slotId, state.productPrefix, msg)
+
+          def debug(msg: String, arg1: AnyRef): Unit =
+            log.debug(s"[{} ({})] $msg", slotId, state.productPrefix, arg1)
```

Comment:
this one seems wrong somehow?

Reply:
I think it's right, we don't check but `msg` should contain another placeholder.